### PR TITLE
Enable COM APIs to log diagnostic and telemetry events and Log Corela…

### PR DESCRIPTION
**Pull request by sachintaMSFT**
…tion Vector Json to existing events

- [ ] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Are you working against an Issue?

-----

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1006)

### CODE REVIEW

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp [hunk 1]
```diff
@@ -11,6 +11,15 @@
 #define AICLI_TraceLoggingStringView(_sv_,_name_) TraceLoggingCountedUtf8String(_sv_.data(), static_cast<ULONG>(_sv_.size()), _name_)
 #define AICLI_TraceLoggingWStringView(_sv_,_name_) TraceLoggingCountedWideString(_sv_.data(), static_cast<ULONG>(_sv_.size()), _name_)
 
+#define AICLI_TraceLoggingWriteActivity(_eventName_,...) TraceLoggingWriteActivity(\
+g_hTraceProvider,\
+_eventName_,\
+GetActivityId(false),\
+nullptr,\
+TraceLoggingCountedString(m_caller.c_str(),  static_cast<ULONG>(m_caller.size()), "Caller"),\
+TraceLoggingPackedFieldEx(GetTelemetryCorelationJsonW().c_str(), static_cast<ULONG>((GetTelemetryCorelationJsonW().size() + 1) * sizeof(wchar_t)), TlgInUNICODESTRING, TlgOutJSON, "CvJson"),\
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:20]: This should use the member, or the function should be changed to return a `const &`.

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp [hunk 2]
```diff
@@ -91,16 +109,54 @@ namespace AppInstaller::Logging
...
+        std::string errors;
+        std::wstring jsonStrW{ jsonStr_view };
+        std::string jsonStr = ConvertToUTF8(jsonStrW.c_str());
+
+        bool result = jsonReader->parse(jsonStr.c_str(),
+            jsonStr.c_str() + jsonStr.size(),
+            pJsonValue.get(),
+            &errors);
+
+        delete jsonReader;
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:128]: This should be a `std::unique_ptr`

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp [hunk 3]
```diff
@@ -91,16 +109,54 @@ namespace AppInstaller::Logging
...
+
+        if (result)
+        {
+            m_telemetryCorelationJsonW = jsonStrW;
+            AICLI_LOG(Core, Info, << "Passed in Corelation Json is valid: " << jsonStr);
+        }
+        else
+        {
+            AICLI_LOG(Core, Info, << "Passed in Corelation Json is invalid: " << jsonStr << "; Error: " << errors);
+            m_telemetryCorelationJsonW = L"{}";
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:138]: The default value should already be this; plus we should make this an error, not a hidden swallow.

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp [hunk 4]
```diff
@@ -133,33 +189,31 @@ namespace AppInstaller::Logging
...
 
-        AICLI_LOG(Core, Info, << "WinGet, version [" << version << "], activity [" << *GetActivityId() << ']');
+        AICLI_LOG(Core, Info, << "WinGet, version [" << version << "], activity [" << *GetActivityId(false) << ']');
         AICLI_LOG(Core, Info, << "OS: " << Runtime::GetOSVersion());
         AICLI_LOG(Core, Info, << "Command line Args: " << Utility::ConvertToUTF8(GetCommandLineW()));
         if (Runtime::IsRunningInPackagedContext())
         {
             AICLI_LOG(Core, Info, << "Package: " << packageVersion);
         }
+        AICLI_LOG(Core, Info, << "IsCOMCall:" << isCOMCall << "; Caller: " << m_caller << "; CallerCj: ");
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/AppInstallerTelemetry.cpp]: Use the same name as the event, not `cj`.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp (hunk 4 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp]: This was mistakenly left over string. removed that. we already log it when setting CvJson.

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp (hunk 2 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:128]: Done.

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp (hunk 1 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:20]: Used member instead.

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp (hunk 2 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:128]: Json::CharReader has virtual methods and is abstract class which cannot be instantiated in make_unique. The derived class OurCharReader isn't abstract whose object pointer newCharReader() is actually returning but it is not declared in Json.h to make unique_ptr to it. How to use std::unique_ptr here ?

Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:128]: Right. Thanks.

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp [hunk 5]
```diff
@@ -11,6 +11,15 @@
 #define AICLI_TraceLoggingStringView(_sv_,_name_) TraceLoggingCountedUtf8String(_sv_.data(), static_cast<ULONG>(_sv_.size()), _name_)
 #define AICLI_TraceLoggingWStringView(_sv_,_name_) TraceLoggingCountedWideString(_sv_.data(), static_cast<ULONG>(_sv_.size()), _name_)
 
+#define AICLI_TraceLoggingWriteActivity(_eventName_,...) TraceLoggingWriteActivity(\
+g_hTraceProvider,\
+_eventName_,\
+GetActivityId(false),\
+nullptr,\
+TraceLoggingCountedString(m_caller.c_str(),  static_cast<ULONG>(m_caller.size()), "Caller"),\
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:19]: `TraceLoggingCountedUtf8String`

File: src/AppInstallerCommonCore/FileLogger.cpp [hunk 6]
```diff
@@ -13,7 +13,7 @@ namespace AppInstaller::Logging
     using namespace std::string_view_literals;
     using namespace std::chrono_literals;
 
-    static constexpr std::string_view s_fileLoggerDefaultFilePrefix = "WinGet-"sv;
+    static constexpr std::string_view s_fileLoggerDefaultFilePrefix = "WPM-"sv;
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/FileLogger.cpp:16]: Ah, I meant change the prefix *for the COM server logs*.

File: src/AppInstallerCommonCore/Public/AppInstallerTelemetry.h [hunk 7]
```diff
@@ -119,6 +125,9 @@ namespace AppInstaller::Logging
 
         bool IsTelemetryEnabled() const noexcept;
 
+        // Check for valid Json string and return a valid Json
+        std::wstring GetTelemetryCorelationJsonW() const;
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Public/AppInstallerTelemetry.h:129]: I still think this would be better as returning a `const std::wstring&`

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCommonCore/FileLogger.cpp (hunk 6 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/FileLogger.cpp:16]: Got it.

File: src/AppInstallerCommonCore/Public/AppInstallerTelemetry.h (hunk 7 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/Public/AppInstallerTelemetry.h:129]: Ok. Yes.

**Review by JohnMcPMS — APPROVED** [review]
I just went ahead and did these to prevent any more back and forth.  If you approve of my changes, it can be merged.

File: src/AppInstallerCommonCore/AppInstallerLogging.cpp [hunk 8]
```diff
@@ -130,6 +130,21 @@ namespace AppInstaller::Logging
         }
     }
 
+    std::string_view DiagnosticLogger::GetLogFileNamePrefix()
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/AppInstallerLogging.cpp:133]: And if I did feel it was necessary to change the prefix of the installer log file, I would want to use something cleaner than this.  If there is no file log, then installer log files have no prefix?

I think that your last two commits should be reverted except for your one change that was unrelated.

If you do want to force the installer log file to have the same prefix, then it needs to be done another way.  The first thing that comes to mind is storing the default as a member on `DiagnosticLogger`.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCommonCore/AppInstallerLogging.cpp (hunk 8 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerLogging.cpp:133]: Undoing the change.

**Review by JohnMcPMS — APPROVED** [review]
I thought I had approved this, sorry.

**Pull request merged**